File: docs/demo/V4.x/TESS_PythonAPI_demo/V4公众号案例代码/停车场/MyNet.py
# -*- coding: utf-8 -*-
import collections
import os
import time
from pathlib import Path
import sys

import PySide2
import pandas as pd

from PySide2.QtCore import *
from PySide2.QtGui import *
from PySide2.QtWidgets import *

import json
import pandas
from functions import tess_function

from Tessng import TessInterface, TessPlugin, NetInterface, PyCustomerNet
from Tessng import tessngPlugin, tessngIFace, m2p, p2m, Online, _DecisionPoint
from Tessng import NetItemType, GraphicsItemPropName




# 用户插件子类，代表用户自定义与路网相关的实现逻辑，继承自MyCustomerNet
class MyNet(PyCustomerNet):
    def __init__(self):
        super(MyNet, self).__init__()
        iface = tessngIFace()
        netiface = iface.netInterface()
    #
    # # 过载的父类方法，当打开网后TESS NG调用此方法
    # #     实现的逻辑是：路网加载后获取路段数，如果路网数为0则调用方法createNet构建路网，之后再次获取路段数，如果大于0则启动仿真
    def afterLoadNet(self):
        # 代表TESS NG的接口
        iface = tessngIFace()
        # 代表TESS NG的路网子接口
        netiface = iface.netInterface()


        # # step1：创建停车区
        parkingRegionLink = [1, 2, 3, 4]    # 指定停车区所在路段
        strPosition = {2: '左侧', 1: '右侧', 0: '车道内'}    # 停车区相对路段的方向
        parkingStallLength = 6    # 停车位长度
        parkingStallWidth = 3    # 停车位宽度
        parkingStallCount = 25    # 每个停车区停车位数量
        positionArray = [1, 2]    # 停车区相对路段的方向
        position2laneNumber = {1: 0, 2: 1}        # 停车区所在车道，右侧停车区在车道0，左侧停车区在车道1
        # 为每个路段创建停车区
        for i in range(len(parkingRegionLink)):
            # 在当前路段的左右两侧分别创建停车区
            for p in positionArray:
                parkingRegionPosition = p
                strParkingRegionPosition = strPosition.get(parkingRegionPosition)
                parkingRegion = Online.ParkingLot.DynaParkingRegion()
                roadId = parkingRegionLink[i]
                laneNumber = position2laneNumber[p]
                parkingRegion.name = f"路段{roadId}{strParkingRegionPosition}停车场"
                parkingRegion.location = 0.5
                parkingRegion.length = parkingStallCount * parkingStallWidth
                parkingRegion.roadId = roadId
                parkingRegion.laneNumber = laneNumber
                parkingRegion.findParkingStallStrategy = 1
                parkingRegion.parkingStallPos = parkingRegionPosition    #  0-车道内;1-车道右侧;2-车道左侧
                parkingRegion.arrangeType = 0     # 0-垂直式;1- 倾斜式-30°;2-倾斜式-45°;3-倾斜式-60°;4-平行式
                # 车位吸引力
                parkingRegion.firstParkingStallAttract = 1
                parkingRegion.middleParkingStallAttract = 1
                parkingRegion.lastParkingStallAttract = 1

                # 停车运动参数
                parkingRegion.menaValue = 0   # 均值
                parkingRegion.variance = 1.0   # 方差
                parkingRegion.parkingSpeed = 5.0   # 泊车速度
                parkingRegion.joinGap = 5.0   # 汇入间隙
                parkingRegion.parkingType = 0    # 0 - 前进->前进;1 - 前进->后退;2 - 后退->前进
                # 运营参数
                parkingRegion.attract = 0   # 吸引力
                parkingRegion.startTime = 0   # 开始时间
                parkingRegion.endTime = 999999   # 结束时间
                parkingRegion.stallLength = parkingStallLength   # 车位长度单位米
                parkingRegion.stallWidth = parkingStallWidth   # 车位宽度单位米
                parkingStalls = []
                # 为当前停车区创建停车位
                for k in range(parkingStallCount):
                    parkingStallData = Online.ParkingLot.DynaParkingStall()
                    parkingStallData.location = parkingRegion.location + parkingRegion.stallWidth * k    # 根据停车区起点位置，计算每个停车位起点位置
                    parkingStallData.parkingStallType = 1    # 停车位类型 // 小客车 / 大客车 / 其他
                    parkingStalls.append(parkingStallData)
                parkingRegion.parkingStalls = parkingStalls
                new_parkingRegion = netiface.createParkingRegion(parkingRegion)
                print(f"创建路段{roadId}{strParkingRegionPosition}停车区{new_parkingRegion.id()}")
        parkingRegionCount = len(netiface.parkingRegions())    # 记录已经创建的停车区数量
        time.sleep(5)

        # # step2：创建停车决策路径
        linkId = 6    # 指定决策点所在路段id
        link = netiface.findLink(linkId)
        distance = link.length() / 3
        name = f"路段{link.id()}上的停车决策点"
        new_parkingDecisionPoint = netiface.createParkingDecisionPoint(link, distance, name)
        print(f"创建路段{link.id()}上的停车决策点")
        for parkingRegion in netiface.parkingRegions():
            parkingRegionId = parkingRegion.id()
            targetParkingRegion = netiface.findParkingRegion(parkingRegionId)
            new_parkingRouting = netiface.createParkingRouting(new_parkingDecisionPoint, targetParkingRegion)
            print(f"创建到停车区{parkingRegionId}的路径{new_parkingRouting.id()}")
        time.sleep(5)

        # step3：平均分配停车区
        update_parkDisInfoList = []
        for parkDisInfo in new_parkingDecisionPoint.parkDisInfoList():
            update_parkDisInfo = Online.ParkingLot.DynaParkDisInfo()
            update_parkDisInfo.pIRouting = parkDisInfo.pIRouting
            update_disVehiclsInfoList = []
            for disVehiclsInfo in parkDisInfo.disVehiclsInfoList:
                update_disVehiclsInfo = Online.ParkingLot.DynaRoutingDisVehicleInfo()
                update_disVehiclsInfo.startTime = disVehiclsInfo.startTime
                update_disVehiclsInfo.endTime = disVehiclsInfo.endTime
                update_vehicleDisDetailList = []
                for vehicleDisDetail in disVehiclsInfo.vehicleDisDetailList:
                    update_vehicleDisDetail = Online.ParkingLot.DynaParkRegionVehicleDisDetail()
                    update_vehicleDisDetail.prop = 1 / parkingRegionCount
                    update_vehicleDisDetail.parkingRegionID = vehicleDisDetail.parkingRegionID
                    update_vehicleDisDetail.parkingRoutingID = vehicleDisDetail.parkingRoutingID
                    update_vehicleDisDetail.parkingSelection = vehicleDisDetail.parkingSelection
                    update_vehicleDisDetail.parkingTimeDisId = vehicleDisDetail.parkingTimeDisId
                    update_vehicleDisDetail.vehicleType = vehicleDisDetail.vehicleType
                    update_vehicleDisDetailList.append(update_vehicleDisDetail)
                update_disVehiclsInfo.vehicleDisDetailList = update_vehicleDisDetailList
                update_disVehiclsInfoList.append(update_disVehiclsInfo)
            update_parkDisInfo.disVehiclsInfoList = update_disVehiclsInfoList
            update_parkDisInfoList.append(update_parkDisInfo)
        result = new_parkingDecisionPoint.updateParkDisInfo(update_parkDisInfoList)
        print(f"为每个停车区平均分配流量, result: {result}")

File: docs/demo/V4.x/TESS_CppAPI_DEMO/1.基础功能/停车场仿真/MyNet.cpp
#include "MyNet.h"

#include <QDebug>

#include "tessinterface.h"
#include "netinterface.h"
#include "ilink.h"
#include "UnitChange.h"
#include "IParkingRegion.h"
#include "IParkingDecisionPoint.h"
#include "IParkingRouting.h"

MyNet::MyNet(){
}

// 加载完路网后的行为
void MyNet::afterLoadNet() {
    // 路网为空时不执行
    if (gpTessInterface->netInterface()->linkCount() == 0) {
        return;
    }
    createParkingRegionsAndRouting();
}

// 创建停车区与停车决策点及路径
void MyNet::createParkingRegionsAndRouting() {
    NetInterface* net = gpTessInterface->netInterface();

    // step1：创建停车区（路段1~4，左右两侧）
    QList<long> parkingRegionLinks = {1, 2, 3, 4};
    QHash<int, QString> posName; posName.insert(1, QString("右侧")); posName.insert(2, QString("左侧")); posName.insert(0, QString("车道内"));
    int parkingStallLengthM = 6;   // 米
    int parkingStallWidthM  = 3;   // 米
    int parkingStallCount   = 25;
    QList<int> positions = {1, 2}; // 右侧/左侧
    QHash<int,int> pos2Lane; pos2Lane.insert(1, 0); pos2Lane.insert(2, 1);

    QList<IParkingRegion*> createdRegions;
    for (long roadId : parkingRegionLinks) {
        for (int p : positions) {
            Online::ParkingLot::DynaParkingRegion pr;
            pr.name = QString("路段%1%2停车场").arg(roadId).arg(posName.value(p));
            pr.location = 0.5; // 起点位置（米）
            pr.length = parkingStallCount * parkingStallWidthM; // 总长度（米）
            pr.roadId = roadId;
            pr.laneNumber = pos2Lane.value(p, 0);
            pr.findParkingStallStrategy = 1;
            pr.parkingStallPos = p; // 0-车道内;1-右侧;2-左侧
            pr.arrangeType = 0;     // 0-垂直式

            // 车位吸引力
            pr.firstParkingStallAttract  = 1;
            pr.middleParkingStallAttract = 1;
            pr.lastParkingStallAttract   = 1;

            // 停车运动参数
            pr.menaValue    = 0;
            pr.variance     = 1.0;
            pr.parkingSpeed = 5.0;
            pr.joinGap      = 5.0;
            pr.parkingType  = 0; // 0-前进->前进

            // 运营参数
            pr.attract   = 0;
            pr.startTime = 0;
            pr.endTime   = 999999;
            // 生成停车位
            QList<Online::ParkingLot::DynaParkingStall> stalls;
            for (int k = 0; k < parkingStallCount; ++k) {
                Online::ParkingLot::DynaParkingStall sd;
                sd.length = parkingStallLengthM; // 米
                sd.width  = parkingStallWidthM;  // 米
                sd.location = pr.location + parkingStallWidthM * k; // 米
                sd.parkingStallType = 1; // 小客车
                stalls << sd;
            }
            pr.parkingStalls = stalls;

            IParkingRegion* r = net->createParkingRegion(pr);
            if (r) {
                createdRegions << r;
                qDebug() << "创建停车区" << r->id();
            }
        }
    }

    // step2：创建停车决策点与到各停车区的路径（位于路段6）
    ILink* link6 = net->findLink(6);
    if (!link6) {
        qDebug() << "未找到路段6，跳过停车决策点与路径创建";
        return;
    }
    qreal distance = link6->length() / 3.0;
    QString dpName = QString("路段%1上的停车决策点").arg(link6->id());
    IParkingDecisionPoint* dp = net->createParkingDecisionPoint(link6, distance, dpName);
    if (!dp) {
        qDebug() << "创建停车决策点失败";
        return;
    }
    for (IParkingRegion* r : createdRegions) {
        IParkingRouting* prt = net->createParkingRouting(dp, r);
        if (prt) {
            qDebug() << "创建到停车区" << r->id() << "的路径" << prt->id();
        }
    }

    // step3：平均分配停车区
    int regionCount = net->parkingRegions().size();
    QList<Online::ParkingLot::DynaParkDisInfo> updateList;
    for (const auto& info : dp->parkDisInfoList()) {
        Online::ParkingLot::DynaParkDisInfo upd;
        upd.pIRouting = info.pIRouting;
        QList<Online::ParkingLot::DynaRoutingDisVehicleInfo> infoList;
        for (const auto& di : info.disVehiclsInfoList) {
            Online::ParkingLot::DynaRoutingDisVehicleInfo ri;
            ri.startTime = di.startTime;
            ri.endTime   = di.endTime;
            QList<Online::ParkingLot::DynaParkRegionVehicleDisDetail> vdList;
            for (const auto& vd : di.vehicleDisDetailList) {
                Online::ParkingLot::DynaParkRegionVehicleDisDetail nv;
                nv.prop = regionCount > 0 ? 1.0 / regionCount : 0.0;
                nv.parkingRegionID  = vd.parkingRegionID;
                nv.parkingRoutingID = vd.parkingRoutingID;
                nv.parkingSelection = vd.parkingSelection;
                nv.parkingTimeDisId = vd.parkingTimeDisId;
                nv.vehicleType      = vd.vehicleType;
                vdList << nv;
            }
            ri.vehicleDisDetailList = vdList;
            infoList << ri;
        }
        upd.disVehiclsInfoList = infoList;
        updateList << upd;
    }
    bool ok = dp->updateParkDisInfo(updateList);
    qDebug() << "为每个停车区平均分配流量, result:" << ok;
}

MyNet::~MyNet(){
}